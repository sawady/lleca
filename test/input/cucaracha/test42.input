
fun par(x : Int) : Bool {
    if x == 0 {
        res := True
    } else {
        res := impar(x - 1)
    } 
    return res
}

fun impar(x : Int) : Bool {
    if x == 0 {
        res := False
    } else {
        res := par(x - 1)
    }
    return res
}

fun probar(x : Int, n : Int) {
    if x <= n {
        if par(x) {
           putChar(112)
        } else {
           putChar(105)
        }
        putNum(x)
        putChar(10)
        probar(x + 1, n)
    }
}

fun main() {
   probar(0, 100)
}
