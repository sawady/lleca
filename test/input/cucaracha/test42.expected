Cons(
  Function(
    par,
    Cons(
      Parameter(
        x,
        Int
      ),
      Nil
    ),
    Bool,
    Cons(
      StmtIf(
        ExprEq(
          ExprVar(
            x
          ),
          ExprConstNum(
            0
          )
        ),
        Cons(
          StmtAssign(
            res,
            ExprConstBool(
              True
            )
          ),
          Nil
        ),
        Just(
          Cons(
            StmtAssign(
              res,
              ExprCall(
                impar,
                Cons(
                  ExprSub(
                    ExprVar(
                      x
                    ),
                    ExprConstNum(
                      1
                    )
                  ),
                  Nil
                )
              )
            ),
            Nil
          )
        )
      ),
      Cons(
        StmtReturn(
          ExprVar(
            res
          )
        ),
        Nil
      )
    )
  ),
  Cons(
    Function(
      impar,
      Cons(
        Parameter(
          x,
          Int
        ),
        Nil
      ),
      Bool,
      Cons(
        StmtIf(
          ExprEq(
            ExprVar(
              x
            ),
            ExprConstNum(
              0
            )
          ),
          Cons(
            StmtAssign(
              res,
              ExprConstBool(
                False
              )
            ),
            Nil
          ),
          Just(
            Cons(
              StmtAssign(
                res,
                ExprCall(
                  par,
                  Cons(
                    ExprSub(
                      ExprVar(
                        x
                      ),
                      ExprConstNum(
                        1
                      )
                    ),
                    Nil
                  )
                )
              ),
              Nil
            )
          )
        ),
        Cons(
          StmtReturn(
            ExprVar(
              res
            )
          ),
          Nil
        )
      )
    ),
    Cons(
      Function(
        probar,
        Cons(
          Parameter(
            x,
            Int
          ),
          Cons(
            Parameter(
              n,
              Int
            ),
            Nil
          )
        ),
        Unit,
        Cons(
          StmtIf(
            ExprLe(
              ExprVar(
                x
              ),
              ExprVar(
                n
              )
            ),
            Cons(
              StmtIf(
                ExprCall(
                  par,
                  Cons(
                    ExprVar(
                      x
                    ),
                    Nil
                  )
                ),
                Cons(
                  StmtCall(
                    putChar,
                    Cons(
                      ExprConstNum(
                        112
                      ),
                      Nil
                    )
                  ),
                  Nil
                ),
                Just(
                  Cons(
                    StmtCall(
                      putChar,
                      Cons(
                        ExprConstNum(
                          105
                        ),
                        Nil
                      )
                    ),
                    Nil
                  )
                )
              ),
              Cons(
                StmtCall(
                  putNum,
                  Cons(
                    ExprVar(
                      x
                    ),
                    Nil
                  )
                ),
                Cons(
                  StmtCall(
                    putChar,
                    Cons(
                      ExprConstNum(
                        10
                      ),
                      Nil
                    )
                  ),
                  Cons(
                    StmtCall(
                      probar,
                      Cons(
                        ExprAdd(
                          ExprVar(
                            x
                          ),
                          ExprConstNum(
                            1
                          )
                        ),
                        Cons(
                          ExprVar(
                            n
                          ),
                          Nil
                        )
                      )
                    ),
                    Nil
                  )
                )
              )
            ),
            Nothing
          ),
          Nil
        )
      ),
      Cons(
        Function(
          main,
          Nil,
          Unit,
          Cons(
            StmtCall(
              probar,
              Cons(
                ExprConstNum(
                  0
                ),
                Cons(
                  ExprConstNum(
                    100
                  ),
                  Nil
                )
              )
            ),
            Nil
          )
        ),
        Nil
      )
    )
  )
)
